### PR TITLE
Add support for REMOTE_USER authentication

### PR DESCRIPTION
**Pull request by abesto**
Useful when Sentry is used behind a company-wide SSO.

### CODE REVIEW

**Comment by slafs** [comment]
Hey! Thanks for the PR. 
I have to tell you I'm really not convinced to ``REMOTE_USER`` support being inside of this image (I kinda regret adding LDAP support to be honest). All those settings regarding User manipulation look very specific to your particular use case. It seems to me that this is a great opportunity for extending the image and making your own. I briefly described how you can do this in README ([link](https://github.com/slafs/sentry-docker#extending-the-image)). Don't you think that it would be even better for you to control this kind of feature in your own image?

I would be more than happy to merge ``SENTRY_PUBLIC`` part if you would open a separate PR.

**Comment by abesto** [comment]
Thanks for explaining your point of view! I feel this is kind of a philosophical difference; I'll try explain what I think. It's totally OK if you still decide not to merge this, I'll then send a separate PR for `SENTRY_PUBLIC`.

Personally I love the idea of being able to just set a bunch of env variables for configuration, and not having to build my own image (12factor and all that). I'd argue that LDAP and REMOTE_USER are both pretty standard solutions for authentication for internal tools. Considering that Sentry is definitely an internal tool, I'd consider lack of support for those in a Sentry package a shortcoming. The beautiful thing is that if I don't set LDAP env vars, nothing LDAP specific happens; same goes for REMOTE_USER.

Sure, building my own image allows simpler customization over the short term, but I imagine it leads to added maintenance in the long term when something changes in upstream (upstream being your image in this case). Personally I prefer a more complex configuration file to potentially tens of forks with slightly different (and so differently buggy) implementations of parts of that complexity. On the other hand I don't have a good feel for the work involved in maintaining a heavily used open-source docker image, so I'll happily defer to your decision :)

**Comment by slafs** [comment]
I see what you mean and I'm little more convinced right now ;). What I'm afraid is that we end up adding more tiny features in this area for people's specific use cases (for example controlling ``is_staff``/``is_superuser`` over some username regex or something like that). What I mean is we probably can't provide one and true solution for this. I don't have much experience with ``RemoteUserBackend`` (I used it only once in some internal project) and can't imagine right now what else people would want in this kind of functionality.

On the other hand there is no huge traffic here regarding pull requests right know, so maybe there's nothing to be afraid of in the first place. :P

**Comment by abesto** [comment]
Right, I can see how it'd be easy to go too far with this kind of over-configurability. Maybe some of the stuff here is already too far. There are basically four components in this PR (plus `SENTRY_PUBLIC`, let me know if I should move that to a separate PR). Which of these do you feel have a place in this package? I'll remove the rest.

 * `SENTRY_USE_REMOTE_USER`
 * `AUTH_REMOTE_USER_HEADER`
 * `is_staff`/`is_superuser`
 * `AUTH_REMOTE_USER_IS_EMAIL`

(I'll also rebase to master in the same go)

**Comment by slafs** [comment]
Yes, please open a seperate PR for ``SENTRY_PUBLIC``. 

As for the features I think we can skip the last two. As I stated before, those feel pretty specific to a certain scenario.

**Comment by abesto** [comment]
All done.

File: sentry_docker_conf.py [hunk 1]
```diff
@@ -240,3 +241,27 @@ def nydus_config(from_env_var):
...
+        # The lazy hack is required because importing RemoteUserMiddleware at load time leads to a circular import
+        # The name is upper camel case because that's the only way to expose values from this config file
+        def build_LAZY_CUSTOM_REMOTE_USER_MIDDLEWARE(header, *args, **kwargs):
+            from django.contrib.auth.middleware import RemoteUserMiddleware
+            class CustomRemoteUserMiddleware(RemoteUserMiddleware):
+               pass
+            CustomRemoteUserMiddleware.header = header
+            return CustomRemoteUserMiddleware(*args, **kwargs)
+        LAZY_CUSTOM_REMOTE_USER_MIDDLEWARE = functools.partial(build_LAZY_CUSTOM_REMOTE_USER_MIDDLEWARE, AUTH_REMOTE_USER_HEADER)
+        MIDDLEWARE_CLASSES += ('sentry_config.LAZY_CUSTOM_REMOTE_USER_MIDDLEWARE',)
```
Comment (slafs) [inline sentry_docker_conf.py]: Just scratching my head on this. How Python/Django will know how to import ``sentry_config.LAZY...`` ?

Comment (abesto) [inline sentry_docker_conf.py]: The Django settings module doesn't have a schema. All the Django config system does is it imports every value from this file that's in `UPPERCASE_UNDERSCORE`. That covers the `LAZY...` part.

As for `sentry_config`, I didn't know either (just saw the module name while trying to implement this). After some digging it turns out that Sentry uses Logan (like this: https://github.com/getsentry/sentry/blob/master/src/sentry/utils/runner.py#L398) to run itself as a stand-alone Django app. And Logan defines the settings module name as `${PROJECT_NAME}_config`: https://github.com/dcramer/logan/blob/master/logan/runner.py#L88

Comment (slafs) [inline sentry_docker_conf.py]: Ahhhh... great! Nice catch!

**Comment by slafs** [comment]
Great PR! Thanks!

**Comment by abesto** [comment]
Thanks for merging!

**Pull request merged**